src/xrpld/app/tx/detail/AMMClawback.cpp

```diff
@@ -0,0 +1,283 @@
+        return tecINTERNAL;  // LCOV_EXCL_LINE
+
+    auto const ammAccount = (*ammSle)[sfAccount];
+    auto const accountSle = sb.read(keylet::account(ammAccount));
+    if (!accountSle)
+        return tecINTERNAL;  // LCOV_EXCL_LINE
+
+    auto const expected = ammHolds(
+        sb,
+        *ammSle,
+        asset,
+        asset2,
+        FreezeHandling::fhZERO_IF_FROZEN,
+        ctx_.journal);
+
+    if (!expected)
+        return expected.error();  // LCOV_EXCL_LINE
+    auto const [amountBalance, amount2Balance, lptAMMBalance] = *expected;
+
+    TER result;
+    STAmount newLPTokenBalance;
+    STAmount amountWithdraw;
+    std::optional<STAmount> amount2Withdraw;
+
+    auto const holdLPtokens = ammLPHolds(sb, *ammSle, holder, j_);
+    if (holdLPtokens == beast::zero)
+        return tecINTERNAL;
+
+    if (!clawAmount)
+    {
+        // Because we are doing a two-asset withdrawal,
+        // tfee is actually not used, so pass tfee as 0.
+        std::tie(result, newLPTokenBalance, amountWithdraw, amount2Withdraw) =
+            AMMWithdraw::equalWithdrawTokens(
+                sb,
+                *ammSle,
+                holder,
+                ammAccount,
+                amountBalance,
+                amount2Balance,
+                lptAMMBalance,
+                holdLPtokens,
+                holdLPtokens,
+                0,
+                WithdrawAll::Yes,
+                ctx_.journal);
+    }
+    else
+        std::tie(result, newLPTokenBalance, amountWithdraw, amount2Withdraw) =
+            equalWithdrawMatchingOneAmount(
+                sb,
+                *ammSle,
+                holder,
+                ammAccount,
+                amountBalance,
+                amount2Balance,
+                lptAMMBalance,
+                *clawAmount);
+
+    if (result != tesSUCCESS)
+        return result;  // LCOV_EXCL_LINE
+
+    auto const res = AMMWithdraw::deleteAMMAccountIfEmpty(
+        sb, ammSle, newLPTokenBalance, asset, asset2, j_);
+    if (!res.second)
+        return res.first;  // LCOV_EXCL_LINE
+
+    JLOG(ctx_.journal.trace())
+        << "AMM Withdraw during AMMClawback: lptoken new balance: "
+        << to_string(newLPTokenBalance.iou())
+        << " old balance: " << to_string(lptAMMBalance.iou());
+
+    auto const ter = rippleCredit(sb, holder, issuer, amountWithdraw, true, j_);
+    if (ter != tesSUCCESS)
+        return ter;  // LCOV_EXCL_LINE
+
+    // if the issuer issues both assets and sets flag tfClawTwoAssets, we
+    // will claw the paired asset as well. We already checked if
+    // tfClawTwoAssets is enabled, the two assets have to be issued by the
+    // same issuer.
+    auto const flags = ctx_.tx.getFlags();
+    if (flags & tfClawTwoAssets)
+
+        return rippleCredit(sb, holder, issuer, *amount2Withdraw, true, j_);
+
+    return tesSUCCESS;
+}
+
+std::tuple<TER, STAmount, STAmount, std::optional<STAmount>>
+AMMClawback::equalWithdrawMatchingOneAmount(
```

Comment:
refactor (optional): I believe this function could be refactored/removed. The statement (lpTokensWithdraw > holdLPtokens) is the same as (!clawAmount). So you could handle this all in applyGuts without another function.

Reply:
then the code will have to declare lpTokensWithdraw and amount2Withdraw in a common place.
And we will have another `if (clawAmount)` to calculate these values. Although in the following check `if (!clawAmount || lpTokensWithdraw > holdLPTokens)`'s body, these values are not used.
```
    STAmount lpTokensWithdraw;
    Number amount2Withdraw;
    if (clawAmount)
    {
        auto frac = Number{*clawAmount} / amountBalance;
        amount2Withdraw = amount2Balance * frac;
        lpTokensWithdraw = toSTAmount(lptAMMBalance.issue(), lptAMMBalance * frac);
    }
    if (!clawAmount || lpTokensWithdraw > holdLPTokens)
           // withdraw all
    else
           // withdraw

```
I prefer to keep the code logic together instead of declaring some variable which can only have value if `if (clawAmount)` is true. And those variables are used in the following `if condition` and `else body`.

Reply:
This is just a simple check with no need for the temp variables:
 
    if (!clawAmount || (lptAMMBalance * *clawAmount / amountBalance) > holdLPTokens)

If you decide to keep the function then makes sense to pass in `holdLPTokens` since it's already fetched.

Reply:
make sense, I'll pass in holdLPTokens